include/rocksdb/utilities/object_registry.h

```diff
@@ -38,6 +39,30 @@ using FactoryFunc =
 // library
 using RegistrarFunc = std::function<int(ObjectLibrary&, const std::string&)>;
 
+// The signature of the function for loading plugins
+// The Plugin struct is passed to the PluginFunc and expected to be validated
+// and filled in by the Plugin code.
+// The Plugin contains the struct of properties associated with this Plugin
+// On success, this function should return 0.  On failure, this function should
+// return non-zero and set the std::string* to an appropriate error message.
+using PluginFunc = std::function<int(Plugin*, std::string*)>;
+
+struct Plugin {
+  // The version of this structure.  If the structure is changed, the version
+  // should be incremented
+  static constexpr int kVersion() { return 1; }
+
+  // The version of the Plugin structure
+  int version = kVersion();
```

Comment:
It seems like we are assuming no ODR[*] violation (or good luck?) in order to check for incompatible versions that happen to succeed in linking. Why would you allow linking to succeed if they're incompatible? (I think I'm proposing to call it PluginV1, possibly with `using Plugin = PluginV1`)

Reply:
This fix was originally to help with exactly the ODR rule.  If the plugin and RocksDB are in separate compilation units (like the Plugin is in a shared library), then it is possible that the Plugin struct is not the same between those units.  This code was put in to help guard against/detect that condition.  If the loadable code (the plugin) detected that the input was different than it expected, it could theoretically do something different.

Out of simplicity of getting this approved, I removed the kVersion/version.